R/shortfall.R

```diff
@@ -0,0 +1,129 @@
+#' and the return of the \emph{actual portfolio}, which contemplates actual transaction prices of executed units net of fees incurred in these transactions.
+#' 
+#' There exist several methods proposed in the literature to calculate the Shortfall.
+#' Each of them firstly considers the difference between the units quantity intended to be traded and the units that were traded, that is the quantity of units that was not traded;
+#' then it further decomposes the above-mentioned portfolios in order to introduce new metrics that provide further post-trade analysis of the trades.
+#' 
+#' The \emph{Complete execution} is the most generic case in which the IS can be calculated, here it is assumed that all the units in the paper portfolio have been executed.
+#' This is the default behaviour of the present function, when no \code{method} argument is provided. See details below.
+#' 
+#' The \emph{Perold's method} allows to express the IS as the sum of an \emph{execution cost} (before-trade), which is the difference between actual and paper prices of the transactions,
+#' and an \emph{opportunity cost}, i.e. the cost of not transacting part of the paper portfolio. 
+#' 
+#' The \emph{Wagner's method} (or "Expanded Implementation Shortfall") allows to take account of the implementation phase of trading decisions,
+#' by separating the period of interest into an \emph{investment period} (time from investment decision to beginning of trading) 
+#' and a \emph{trading period} (time from beginning of trading to end of trading). A \emph{price change} in between is obtained 
+#' and it becomes in turn an input to Perold's method, allowing us to reformulate the IS is terms of the two periods above and thus to identify 
+#' a \emph{delay related cost} and a \emph{trading related cost}, in addition to (a remainder of) Perold's opportunity cost and the transactions fees.
+#' A particular istance of Wagner's method is the so called \emph{Market Activity Implementation Shortfall}, which takes place when the trades paper prices are not provided (assumed to be unknown) and thus the corresponding arrival prices are used instead.
+#' Furthermore, Wagner's method allows for an additional decomposition of the "Delay related cost" into the \emph{opportunity delay cost} component and the \emph{trading delay cost} component.
+#' 
+#' 
+#' @param Portfolio A portfolio name that points to a portfolio object structured with initPortf()
+#' @param Symbol A string identifying the symbol to calculate the Shortfall for
+#' @param PaQty The total number of units (such as shares or contracts) intended to trade
+#' @param PaPriceStart Decision price at the beginning of trading
+#' @param PaPriceEnd Decision future price at the end of trading
+#' @param method String specifying which method to use for The Implementation Shortfall calculation. One of 'Perold' or 'Wagner'
```

Comment:
Hi @jaymon0703, I have noticed you updated the issue with a very comprehensive presentation of the IS models. There you proposed to implement the function using the parameter `model=c("Market","Perold","Wagner")`, before this was online I used `method` in a very similar fashion, would it be better to change it? Also, you proposed to default this parameter to the Market Activity Implementation Shortfall, but I am not sure I fully understand, would you please provide further comments on this? Currently, the parameter defaults to what Kissell calls the 'Complete execution' case, which is the only case in which the shortfall is calculated assuming that all the intended trade units (`PaQty`) were executed. Strictly speaking, I would not consider it to be a method or model per se and this is why I put it before the "proper" methods/models.
In addition, I used the fact that it is a particular case of Wagner's method (I left a note at line 114) as it seems to me the easier way to implement it. Its difference is only in the meaning of the output. In a certain sense, it is the default behavior of the function when `method='Wagner'`, because when no `PaPriceStart` is provided, the `ArrPrice` (what you named `p_0`) is used by default. What do you think if I stress this aspect in the documentation or we make it produce a message when this happens?

Reply:
Thanks for the PR @vi-to 

I think a good general architecture can be:
If no decision price is specified, then use Market Activity IS (equation 3.8) with arrPrice as you have specified. The key difference between Market Activity IS and the others is that no decision price is specified. We may want to add arrPrice as a separate parameter, since arrival price is really the middle of the bid-ask spread at the time the order enters the orderbook. If this is not specified, then use first transaction price. Also, Market Activity IS is exactly Expanded (ie. Wagner) IS without the Delay Cost (ie. first term S(P0-Pd)).

If decision price is present, then the other 2 methods (Wagner & Perold) come into play.

I think method=c("Market", "Perold", "Wagner") would be useful, and we should make the methods explicit.

Lastly, it may be worth including an example, such as 

```suggestion
data(amzn)
txns <- amzn.trades[which(amzn.trades[,2] > 0),]
try(rm("portfolio.testport","account.testport",pos=.blotter), silent = TRUE)
stock.str='amzn' # what are we trying it on
currency('USD')
stock(stock.str,currency='USD',multiplier=1)
initPortf("testport", symbols=stock.str)
initAcct("testport","testport", symbols=stock.str)
addtxns <- addTxns("testport","amzn",txns)
updatePortf("testport")
p = getPortfolio("testport") # make a local copy of the portfolio object
a = getAccount("testport") # make a local copy of the account object
p$symbols$amzn$txn

implementationShortfall("testport","amzn",3300,127,128, method = 'Market')
implementationShortfall("testport","amzn",3300,127,128,method = 'Perold')
implementationShortfall("testport","amzn",3300,127,128,method = 'Wagner')
```

The respective Implementation Shortfall metrics for Perold and Wagner should be identical when used with similar inputs.

Reply:
Found two errors in current formulations (ironically alternatives were 'correct'), committing the corrections.

Reply:
Perfect, thanks @vi-to

Reply:
Hu @jaymon0703, the last commit adds a new flow for methods where `method='Market'` is explicitly accounted for. The default `method` has not changed yet, as I feel that it may useful to have for "quick calcs" and that otherwise we won't have a chance to provide the user with `PaRet` and `AcRet` elsewhere. However, an option is to make it a "method". What do you think about this?

Reply:
Thanks Vito

I have had a chance to review the PR and have a few observations and suggestions.

1. It would be useful to specify all the methods in the functions signature, such as method = c("Complete", "Market", "Perold", "Wagner"). If no value is specified, the default will be Complete. I agree, having Complete as an available method will be useful.

2. This speaks to your question in issue #81 and i think its worth addressing sooner rather than later in case others are looking for guidelines for contributing to blotter. Happy to work with you on a contributing guidelines doc along the lines of https://help.github.com/en/articles/setting-guidelines-for-repository-contributors. Nevertheless, we tend to have a preference for camelCase when setting variable or param names. So for example ArrPrice could rather be arrPrice and PaPriceStart could rather be priceStart, etc.

3. Your calculation for Complete Execution looks incorrect. Pn is not relevant when all trades intended for execution are ultimately executed. I have updated the issue #54 with the formula we should be using for Complete Execution. In AcRet replace PaPriceEnd with the average execution price. Paper return is irrelevant, because the entire quantity was traded.

4. Fees are added as negative amounts in AddTxns, so we will have to subtract the fees variable in order to add it to the Implementation Shortfall measure. So in line 89, perhaps `Fees <- -1 * sum(txns$Txn.Fees)`

5. Can we rename the file impShortfall and likewise the function name? Shortfall can be confused with other functions in other packages such as Expected Shortfall in PerformanceAnalytics so we should be a bit more explicit there. In the long run we may want to consolidate all TCA models into one TCA.R file, but for now impShortfall.R will do.

6. I think it would be useful to add a column to the dataframe output for the type of method used. For example, if someone loops through all methods they would want to see which method relates to which output.

7. I have set up dummy trades to mirror the examples in the book and i think these could be useful examples in the documentation...once the change for fees and the correction for "Complete" are implemented i think we will be in a position to merge with the tca branch in blotter. If you dont mind i would like to add a bit more context in the documentation, once it is merged.

Thanks again for the PR.

Tests to replicate the examples in the book...

```# set up test_txns assuming all 5k shares traded
test_txns <- xts(cbind(rep(10:11,5),rep(500,10),TxnFees = rep(-10,10)),
                 seq.POSIXt(as.POSIXct("2000-01-01 09:00:00"), as.POSIXct("2000-01-01 18:00:00"), by = "hours"))
colnames(test_txns) <- c("TxnPrice","TxnQty","TxnFees")

stock.str='tca_test' # what are we trying it on
currency('USD')
stock(stock.str,currency='USD',multiplier=1)
suppressWarnings(rm("account.testport","portfolio.testport",pos=.blotter))
initPortf("testport", symbols=stock.str)
initAcct("testport","testport", symbols=stock.str)
addtxns <- addTxns("testport","tca_test",test_txns)
updatePortf("testport")
p = getPortfolio("testport") # make a local copy of the portfolio object
a = getAccount("testport") # make a local copy of the account object
p$symbols$tca_test$txn

### Complete Execution
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice=10)

#### Market
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice = 10, method='Market')

# set up test_txns assuming assuming only 4k shares traded
test_txns <- test_txns[-c(2:3),]
suppressWarnings(rm("account.testport","portfolio.testport",pos=.blotter))
initPortf("testport", symbols=stock.str)
initAcct("testport","testport", symbols=stock.str)
addtxns <- addTxns("testport","tca_test",test_txns)
updatePortf("testport")
p = getPortfolio("testport") # make a local copy of the portfolio object
a = getAccount("testport") # make a local copy of the account object
p$symbols$tca_test$txn

### Perold
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice=10, method='Perold')

### Wagner
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice=10.25, method='Wagner')
```

Reply:
Thank you @jaymon0703 for your comprehensive review.

I will first address (1), (4) and (6). 

Then (7), (2), (5) and docs update also with examples in (8), possibly in this order. 

As for (3):
while the alternative expression is correct, I chose to implement it fully to be consistent with the others (at the moment the function is completely independent of `p_avg`) and provide its two underneath quantities. When the provided `PaQty` equals `sum(txns$Txn.Qty)`, they are equivalent. It may happen however that one provides an inconsistent `PaQty`. Furthermore, once `PaPriceEnd` default value will be in place, there will be no need to provide it, especially when `method='Complete'` (in accordance with (7)).
If you think the full implementation is valuable, here I propose to force `PaQty` to `sum(txns$Txn.Qty)` and add a warning or alternatively to add a stop if they differ.

About (8):
what you are proposing is great, indeed in doing so I encourage you to add yourself as an author of the function if you want to.
Related with the docs is that, once we have finished the implementation, I was thinking to factor out `sum(txns$Txn.Qty)` to simply the expressions and rename it to avoid possible confusions with other `blotter` produced outputs (maybe `TotTxnQty`).

Reply:
Ok thanks @vi-to, looks like your latest commit resolves 1, 4 and 6. Once 7, 2, 5 and 8 are resolved we can merge this and move on.

Reply:
@jaymon0703 I am about to commit the function name change. I have corrected spacing as well, at this point do you want me to change the `.R` name too?